### PR TITLE
Identify layer features Sample 

### PR DESCRIPTION
**Pull request by prupani-7**
## Description

PR to add a new Kotlin sample _"Identify layer features"_ in `Search and Query` category.

## Links and Data

Sample Epic: `runtime/kotlin/issues/2934`

## What To Review

-  Review the code to make sure it is easy to follow like other samples on Android
- `README.md` and `README.metadata.json` files

## How to Test

Run the sample on the sample viewer or the repo.

### CODE REVIEW

File: identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/components/ComposeMapView.kt [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            mapView.apply {
+                map = mapViewState.arcGISMap
+                setViewpoint(mapViewState.viewpoint)
+            }
+        }
+    )
+
+    // launch coroutine functions in the composition's CoroutineContext
+    LaunchedEffect(Unit) {
+        launch {
```
Comment (shubham7109) [inline identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/components/ComposeMapView.kt:62]: Remove `launch`
```suggestion
    LaunchedEffect(Unit) {
```

File: identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/components/ComposeMapView.kt [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            mapView.onSingleTapConfirmed.collect {
+                // call identifyLayers when a tap event occurs
+                val identifyResult = mapView.identifyLayers(it.screenCoordinate, 12.0, false, 10)
```
Comment (shubham7109) [inline identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/components/ComposeMapView.kt:65]: This sample showcases the identify call in the API, let's try to use named arguments here to show which type of identify call is being used here:
```suggestion
                // call identifyLayers when a tap event occurs
                val identifyResult = mapView.identifyLayers(
                    screenCoordinate = it.screenCoordinate,
                    tolerance = 12.0,
                    returnPopupsOnly = false,
                    maximumResults = 10
                )
```

File: identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/screens/MainScreen.kt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                        .fillMaxSize()
+                        .weight(1f),
+                    mapViewModel = mapViewModel
+                )
+                // Bottom text to display the identify results
+                Row(
+                    modifier = Modifier
+                        .padding(12.dp)
+                        .fillMaxWidth()
+                ) {
```
Comment (shubham7109) [inline identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/screens/MainScreen.kt:67]: The bottom row is causing a resize to both the `Row` and `ComposeMapView`, this looks janky without any transitions. A quick fix would be to set the Modifier to use `animateContentSize()`, which should help resize both elements with transitions. 

```suggestion
                ComposeMapView(
                    modifier = Modifier
                        .fillMaxSize()
                        .weight(1f)
                        .animateContentSize(),
                    mapViewModel = mapViewModel
                )
                // Bottom text to display the identify results
                Row(
                    modifier = Modifier
                        .padding(12.dp)
                        .fillMaxWidth()
                        .animateContentSize()
                ) {
```

File: identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/components/MapViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+import com.arcgismaps.mapping.layers.FeatureLayer.Companion.createWithFeatureTable
+import com.arcgismaps.mapping.view.IdentifyLayerResult
+import com.esri.arcgismaps.sample.identifylayerfeatures.R
+import com.esri.arcgismaps.sample.sampleslib.components.MessageDialogViewModel
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.launch
+
+class MapViewModel(
+    private val application: Application,
```
Comment (shubham7109) [inline identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/components/MapViewModel.kt:38]: `application` is only used in the `init` block
```suggestion
    application: Application,
```

File: identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/components/MapViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    private val sampleCoroutineScope: CoroutineScope
+) : AndroidViewModel(application) {
+    // set the MapView mutable stateflow
+    val mapViewState = MutableStateFlow(MapViewState())
+
+    // create a ViewModel to handle dialog interactions
+    val messageDialogVM: MessageDialogViewModel = MessageDialogViewModel()
+
+    // string text to display the identify layer results
+    val bottomTextBanner = mutableStateOf(String())
```
Comment (shubham7109) [inline identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/components/MapViewModel.kt:48]: Let's set a default text message to the banner: 

```suggestion
    val bottomTextBanner = mutableStateOf("Tap on the map to identify feature layers")
```

File: identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/components/MapViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        mapViewState.value.arcGISMap = map
+    }
+
+    /**
+     * Identify the feature layer results and display the resulting information
+     */
+    fun handleIdentifyResult(result: Result<List<IdentifyLayerResult>>) {
+        sampleCoroutineScope.launch {
+            result.onSuccess { identifyResultList ->
+                bottomTextBanner.value = ""
```
Comment (shubham7109) [inline identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/components/MapViewModel.kt:86]: Don't think this is needed here, the `bottomTextBanner` should be set if `totalCount > 0`, else, set the text to be "Number of elements found: N/A"
```suggestion
```

File: identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/components/MapViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+                    totalCount += geoElementsCount
+                }
+                // if any elements were found show the results, else notify user that no elements were found
+                if (totalCount > 0) {
+                    bottomTextBanner.value = "Number of elements found:\n${message}"
+                } else {
+                    messageDialogVM.showMessageDialog(
+                        title = "No element found",
+                        description = message.toString()
+                    )
```
Comment (shubham7109) [inline identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/components/MapViewModel.kt:107]: The `message` would be empty if no elements are found. Maybe use a custom description for the dialog? 

```suggestion
                    messageDialogVM.showMessageDialog(
                        title = "No element found",
                        description = "Tap an area on the map with visible features"
                    )
```

File: identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/components/MapViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+            subLayerGeoElementCount += geoElementsCountFromResult(sublayerResult)
+        }
+        return subLayerGeoElementCount + result.geoElements.size
+    }
+}
+
+/**
+ * Data class that represents the MapView state
+ */
+data class MapViewState( // This would change based on each sample implementation
```
Comment (shubham7109) [inline identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/components/MapViewModel.kt:137]: Remove comment

```suggestion
data class MapViewState( 
```

File: identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/components/MapViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    var arcGISMap: ArcGISMap = ArcGISMap(BasemapStyle.ArcGISNavigationNight),
+    var viewpoint: Viewpoint = Viewpoint(
+        Point(
+            -10977012.785807, 4514257.550369,
+            SpatialReference(wkid = 3857)
+        ), 68015210.0
+    )
```
Comment (shubham7109) [inline identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/components/MapViewModel.kt:144]: Couple nested constructors being accessed here, makes it somewhat confusing on how the viewpoint is being set. Maybe we could format it using named arguments to help better clarify the parameters: 

```suggestion
    var viewpoint: Viewpoint = Viewpoint(
        center = Point(
            x = -10977012.785807,
            y = 4514257.550369,
            spatialReference = SpatialReference(wkid = 3857)
        ),
        scale = 68015210.0
    )
```

**Review by shubham7109 — CHANGES_REQUESTED** [review]
Nice sample @prupani-7! Couple comments here:

**Comment by prupani-7** [comment]
@shubham7109 Thanks for the review suggestions! Applied. please re-review?

File: identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/components/MapViewModel.kt [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+                    // add new line character if not the final element in array
+                    if (identifyLayerResult != identifyResultList[identifyResultList.size - 1]) {
+                        message.append("\n")
+                    }
+                    totalCount += geoElementsCount
+                }
+                // if any elements were found show the results, else notify user that no elements were found
+                if (totalCount > 0) {
+                    bottomTextBanner.value = "Number of elements found:\n${message}"
+                } else {
```
Comment (shubham7109) [inline identify-layer-features/src/main/java/com/esri/arcgismaps/sample/identifylayerfeatures/components/MapViewModel.kt:102]: Could you apply this suggesstion: 
https://github.com/Esri/arcgis-maps-sdk-kotlin-samples/pull/140#discussion_r1375016787

```suggestion
                } else {
                   bottomTextBanner.value = "Number of elements found: N/A"
```

**Review by shubham7109 — APPROVED** [review]
Looks good to approve 👍

**Review by rchintapalli — APPROVED** [review]
Not sure if it was part of original sample design or not, but selecting/highlighting the identified features also gives better visual experience.

Looks good to me otherwise

**Comment by prupani-7** [comment]
@rchintapalli sample design doesn't do feature selecting/highlighting. So leaving this as is. Thanks for your review.

**Pull request merged**